src/libs/websocket/data.cpp

```diff
@@ -0,0 +1,1053 @@
+	                              "set_machine_state",
+	                              "set_order_delivered",
+	                              "set_robot_maintenance",
+	                              "set_teamname",
+	                              "reset_machine_by_team"};
+
+	for (const std::string &schema_name : schema_names) {
+		rapidjson::SchemaDocument *sd =
+		  load_schema(base_path + "/src/libs/websocket/message_schemas/" + schema_name + ".json");
+		if (sd) {
+			command_schema_map[schema_name] = sd;
+		} else {
+			throw Exception("No schema file could be found for '%s'", schema_name.c_str());
+		}
+	}
+}
+
+/**
+ * @brief Read a JSON schema file and return pointer to a rapidjson SchemaDocument
+ * 
+ * @param path 
+ * @return rapidjson::SchemaDocument 
+ */
+rapidjson::SchemaDocument *
+Data::load_schema(std::string path)
+{
+	rapidjson::Document d;
+
+	FILE *fp = std::fopen(path.c_str(), "rb");
+	if (fp == 0) {
+		logger_->log_error("Websocket", "couldn't load JSON schema file");
+		return nullptr;
+	}
+	char                      readBuffer[65536];
+	rapidjson::FileReadStream is(fp, readBuffer, sizeof(readBuffer));
+	if (d.ParseStream(is).HasParseError()) {
+		logger_->log_error("Websocket", "couldn't parse JSON schema file");
+		return nullptr;
+	}
+	rapidjson::SchemaDocument *schema = new rapidjson::SchemaDocument(d);
+	return schema;
+}
+
+/**
+ * @brief get first element from log queue
+ * 
+ *  This thread-safe function returns the first element from the log message queue and removes it.
+ * 
+ * @return std::string first element from log queue
+ */
+std::string
+Data::log_pop()
+{
+	const std::lock_guard<std::mutex> lock(log_mu);
+	std::string                       log = logs.front();
+	logs.pop();
+	return log;
+}
+
+/**
+ * @brief add element to log queue
+ * 
+ * This thread-safe function adds an element to the log message queue. 
+ * 
+ * @param log element (std::string) to be added
+ */
+void
+Data::log_push(std::string log)
+{
+	const std::lock_guard<std::mutex> lock(log_mu);
+	logs.push(log);
+	log_cv.notify_one();
+}
+
+/**
+ * @brief add JSON element to log queue
+ * 
+ * This thread-safe function adds a JSON element to the log message queue. 
+ * 
+ * @param d element (rapidjson::Document) to be added
+ */
+void
+Data::log_push(rapidjson::Document &d)
+{
+	const std::lock_guard<std::mutex>          lock(log_mu);
+	rapidjson::StringBuffer                    buffer;
+	rapidjson::Writer<rapidjson::StringBuffer> writer(buffer);
+	d.Accept(writer);
+
+	logs.push(buffer.GetString());
+	log_cv.notify_one();
+}
+
+/**
+ * @brief check if log queue is empty
+ * 
+ * @return true log queue is empty
+ * @return false log queue is not empty
+ */
+bool
+Data::log_empty()
+{
+	return logs.empty();
+}
+
+/**
+ * @brief blocks calling thread until log queue is not-empty
+ * 
+ */
+void
+Data::log_wait()
+{
+	std::unique_lock<std::mutex> lock(log_mu);
+	while (log_empty()) //loop to avoid spurious wake-ups
```

Comment:
Actually this is not busy waiting at all, the thread is waiting for a mutex to be released. However, when researching mutexes in C++, some reference (would have to search which one again) recommends adding a loop that checks for the actual wait up condition too, because apparently spurious wake-ups might still occur. So it looks like busy waiting but isn't.